Added parameter to control if the service starts immediately when installing

fhofmann opened the pull request:
Added a parameter to control the startImmediately behaviour of the win32 service when installing the service.
This is useful when installing, then configuring something, and then starting the service after this configuration step is done.
Default is the same as before.

avisonmaher commented:
How weird. I've just been working on something to try and achieve the same outcome.

My solution was just adding another action so you would do "MyService action:delayedinstall"  to get it to just install the service. I see you've gone down the parameter route.

For yours, would you have to do "MyService action:install startimmediately:true" ?

fhofmann commented:
Yeah, true would be the default, so action:install startimmediately:false would be the same as your action:delayedinstall.
One more thing I thought about, the service will still be set to start automatically (for instance when restarting the system - i think).
That could be another parameter, but at least for our use case this isn't needed.

PeterKottas commented:
Hi guys, just looking at this now. It definitely should be a part of this lib if you found a sensible use-case for it. It's just a matter of choosing the best syntax for it as both this and #56 have pretty much the same functionality. I am slightly inclining towards the solution by @avisonmaher as it seems a bit more verbose and won't require and extra parameter. What do you think guys? Any reasons come to mind why we should use one or another?

fhofmann commented:
Hey, when I implemented this, I just looked at the underlying calls to the service and saw that there was an extra paramerter, that was always set to true, so I added one to forward this functionality. I don't really see a reason to favor one over the other. They both provide the same functionality, "delayedinstall" is maybe not the best name, but tbh I can't really think of a better one.

avisonmaher commented:
I don't mind, I'd agree "delayedinstall" isn't the best name. Taking inspiration from https://ss64.com/nt/sc.html it could be changed to "create". 
I think @fhofmann 's solution is a quicker and tidier way right now but does require typing more to get the service installed. I've created a new option on an existing parameter however it has created duplicated code so probably requires a bit more work to tidy it up with influences from #55 - which I've been thinking about but haven't gotten too.

PeterKottas commented on Source/PeterKottas.DotNetCore.WindowsService/HostConfiguration.cs:
```diff
@@ -30,6 +30,8 @@ public HostConfiguration()
 
         public string DisplayName { get; set; }
 
+        public bool? StartImmediately { get; set; }
```
Could you please change this to not nullable and add default property value?

PeterKottas commented on Source/PeterKottas.DotNetCore.WindowsService/ServiceRunner.cs:
```diff
@@ -203,7 +215,7 @@ private static void Install(HostConfiguration<SERVICE> config, ServiceController
                     cred,
                     autoStart: true,
-                    startImmediately: true,
+                    startImmediately: config.StartImmediately ?? true,
```
if it's nullable it should be config.StartImmediately.HasValue ? config.StartImmediately.Value : true instead. But I think it's better to make it non nullable like mentioned in my previous comment.

PeterKottas commented on Source/PeterKottas.DotNetCore.WindowsService/ServiceRunner.cs:
```diff
@@ -130,36 +142,36 @@ public static int Run(Action<HostConfigurator<SERVICE>> runAction)
+                ConfigureService(innerConfig);
 
-				return 0;
+                return 0;
```
There seems to be a lot of false positive diffs because of formatting changes. Just some extra white spaces I guess. Could you please reformat this doc?

PeterKottas commented:
Ok I'll be merging this one in guys. Just a few small issues with the code so please @fhofmann could you take a look? I'll be merging in and releasing right after this is done.

fhofmann commented:
Thanks for the review. 
Okay, should be done. Do I have to do a new pull request or do you see the changes automatically?
Cheers
Frieder

PeterKottas commented:
Cheers mate, all good. There's still some formatting issues even after this commit but screw that. I'll fix those myself and this PR is ok to go in as far as I am concerned.

avisonmaher commented:
I've tried this by adding both:

MyWindowsService startimmediately:false action:install   (as discussed in here)
MyWindowsService start-immediately:false action:install  (as referenced on the documentation)

and tried swapping the parameters around:

MyWindowsService action:install startimmediately:false
MyWindowsService action:install start-immediately:false 

But it isn't working, it is installing the service and running it straight away.

fhofmann commented:
For me it works fine:

 action:install start-immediately:true
-> service running
action:install start-immediately:false
-> service is there, but is stopped.

fhofmann commented:
One thing I noticed, the message being displayed might be miss leading when using action:install and start-imediately:false

"Successfully registered and started service" will be written to the console.
